### ISSUE TITLE
Add changes generated by pre-commit

### ISSUE THREAD

**Issue by henniss** [pre-commit/pre-commit#879]
Loosely related to #806 . 

I sometimes find myself doing the following: 

* Edit some files. 
* Run `git add -p` and stage only some of my edits. 
* Run `git commit -m "Some message"`. 
  * pre-commit stashes all the unstaged changes,
  * runs all hooks, which change the working tree,
  * pops the stashed edits,
  * and rejects the commit.

The changes generated by the pre-commit hooks are now intermixed with the unstaged edits. I can't just run `git add -u && !!`, as that will stage too much. Instead I need to run `git add -p` and select only the unstaged changes generated by my hooks. 

Could pre-commit generate a patch of changes it generates and store that patch in `.git`? Then I could do something like the following:

```bash
git add -p
git commit -m "Some message."
cat .git/pre-commit/patch
git apply --index .git/pre-commit/patch && !-2
```

Or is there already a way to do this?

**Comment by asottile** [comment]
as you've learned from #806 and #747 -- pre-commit will not modify your staging area since this is judged as an unsafe operation

as for generating a patch, there is one which is stored in `~/.cache/pre-commit` which represents the parts pre-commit unstages.  You're free to reverse-apply that, then staged, then commit, then apply the patch again and you'll be back to your desired state.

Let me know if this works for you!

Here's a sample workflow for that:

```diff
$ git diff --staged
diff --git a/t b/t
index f00c965..dc76404 100644
--- a/t
+++ b/t
@@ -1,6 +1,6 @@
 1
 2
-3
+3 # trailing whitespace 
 4
 5
 6
$ git diff
diff --git a/t b/t
index dc76404..8e6f928 100644
--- a/t
+++ b/t
@@ -4,7 +4,7 @@
 4
 5
 6
-7
+7 # don't stage me
 8
 9
 10
```

And when I commit:

```
$ git commit -m "hi"
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /Users/asottile/.cache/pre-commit/patch1544222126.
Trim Trailing Whitespace.................................................Failed
hookid: trailing-whitespace

Files were modified by this hook. Additional output:

Fixing t

Fix End of Files.........................................................Passed
Check Yaml...........................................(no files to check)Skipped
Check for added large files..............................................Passed
[INFO] Restored changes from /Users/asottile/.cache/pre-commit/patch1544222126.
```

Now I can remove the parts I didn't want to commit:

```
$ git apply --reverse /Users/asottile/.cache/pre-commit/patch1544222126
$ git add -u
$ git diff --staged
diff --git a/t b/t
index f00c965..50cfb57 100644
--- a/t
+++ b/t
@@ -1,6 +1,6 @@
 1
 2
-3
+3 # trailing whitespace
 4
 5
 6
```

And then commit and then restore:

```diff
$ git commit -m "wat"
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Passed
Check Yaml...........................................(no files to check)Skipped
Check for added large files..............................................Passed
[master 13f3873] wat
 1 file changed, 1 insertion(+), 1 deletion(-)
$ git apply /Users/asottile/.cache/pre-commit/patch1544222126
$ git diff
diff --git a/t b/t
index 50cfb57..6900639 100644
--- a/t
+++ b/t
@@ -4,7 +4,7 @@
 4
 5
 6
-7
+7 # don't stage me
 8
 9
 10
```

**Comment by henniss** [comment]
Thanks, that's perfect. Somehow I totally overlooked the info messages that referenced the cache.

**Comment by asottile** [comment]
awesome!  Thanks for the issue

**Comment by blueyed** [comment]
I often run into this myself..

A workaround is also to use `git stash -k` (`--keep-index`) before committing, so any changes from pre-commit can be added using `git add -u`.

But it needs to be done before.

I would like if pre-commit could just not restore the changes from the patch (optionally), allowing for `git add -u`, and then restoring it manually - a command to do so could be given for copy'n'paste.

**Comment by Abbyyan** [comment]
It's weird that i didn't find the `~/.cache/.cache/pre-commit`。 How can i achieve this ? My git version is `git version 2.17.0` and `pre-commit` file just like this :
```shell
git diff --cached |  /home/lenajin/tools/clang-format/clang-format-diff.py -i -p1
``` 
Thanks a lot .

**Comment by asottile** [comment]
hello @Abbyyan -- this issue is about something completely different, and I suspect you're confusing `.git/hooks/pre-commit` (the git hook named `pre-commit`) with pre-commit.com (the framework housed in this repository)

if you have additional questions please open a new issue

**Comment by brunolnetto** [comment]
This is a short issue with quick answer. On it, @asottile and @blueyed provide some workflow good practices to operate manually on pre-commit along its changes. My question is on pre-commit as an gh-action. I saw this work (https://pre-commit.ci/), but as a Junior developer, I lack the skills to set a yml-only to lint (and not just lint-check) the incoming commits.

IMHO a dev-person should not be always actively aware of pre-commit changes. In fact, it should apply the changes by the above-mentioned  action.
